confidential-data-hub/kms/src/plugins/ehsm/client.rs

```diff
@@ -0,0 +1,228 @@
+            &provider_settings.endpoint,
+        )
+    }
+
+    /// Export the [`ProviderSettings`] of the current client. This function is to be used
+    /// in the encryptor side. The [`ProviderSettings`] will be used to initial a client
+    /// in the decryptor side.
+    pub fn export_provider_settings(&self) -> Result<ProviderSettings> {
+        let provider_settings = EhsmProviderSettings {
+            app_id: self.client.appid.clone(),
+            endpoint: self.client.base_url.clone(),
+        };
+
+        let provider_settings = serde_json::to_value(provider_settings)
+            .map_err(|e| Error::EhsmKmsError(format!("serialize ProviderSettings failed: {e}")))?
+            .as_object()
+            .expect("must be an object")
+            .to_owned();
+
+        Ok(provider_settings)
+    }
+}
+
+#[async_trait]
+impl Encrypter for EhsmKmsClient {
+    async fn encrypt(&mut self, data: &[u8], key_id: &str) -> Result<(Vec<u8>, Annotations)> {
+        let ciphertext = self.client.encrypt(
+                &key_id.to_owned(),
+                &STANDARD.encode(data),
+                None
+            )
+            .await
+            .map_err(|e| Error::EhsmKmsError(format!("EHSM-KMS encrypt failed: {e}")))?;
+            
+        let annotations = Annotations::new();
+
+        Ok((ciphertext.into(), annotations))
+    }
+}
+
+#[async_trait]
+impl Decrypter for EhsmKmsClient {
+    async fn decrypt(
+        &mut self,
+        ciphertext: &[u8],
+        key_id: &str,
+        _annotations: &Annotations,
+    ) -> Result<Vec<u8>> {
+        let plaintext_b64 = self.client.decrypt(
+                &key_id.to_owned(),
+                &std::str::from_utf8(&ciphertext).map_err(|e| Error::EhsmKmsError(format!("decrypt &[u8] to &str failed: {e}")))?,
+                None
+            )
+            .await
+            .map_err(|e| Error::EhsmKmsError(format!("EHSM-KMS decrypt failed: {e}")))?;
+        let plaintext = STANDARD.decode(plaintext_b64)
+            .map_err(|e| Error::EhsmKmsError(format!("decode plaintext for decryption failed: {e}")))?;
+
+        Ok(plaintext)
+    }
+}
+
+impl EhsmKmsClient {
+    pub async fn create_key(&mut self, key_spec: &str) -> Result<String> {
+        let origin = "EH_INTERNAL_KEY";
+        let keyusage = "EH_KEYUSAGE_ENCRYPT_DECRYPT";
+        let key_id = self.client.create_key(&key_spec, &origin, &keyusage)
+            .await
+            .map_err(|e| Error::EhsmKmsError(format!("EHSM-KMS create key failed: {e}")))?;
+
+        Ok(key_id)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rstest::rstest;
+    use serde_json::json;
+
+    use crate::{plugins::ehsm::client::EhsmKmsClient, Decrypter, Encrypter};
+
+    #[ignore]
+    #[tokio::test]
+    async fn test_create_key() {
+        let key_spec = "EH_AES_GCM_256";
+        let provider_settings = json!({
+            "app_id": "86f0e9fe-7f05-4110-9f65-a224ddee1233",
+            "endpoint": "https://172.16.1.1:9002",
```

Comment:
Is this a real instance running somewhere? Is it guaranteed to be running?

Reply:
Sure! it is an eHSM-KMS service that I have deployed on my local machine. And all tests pass successfully.
![AA31A454-D5A6-463D-AC06-3864653EF76D](https://github.com/confidential-containers/guest-components/assets/20621043/e950ad66-efd8-4687-9667-bee40257b247)

Reply:
By the way, it's hard to run these tests on CoCo CI, since eHSM-KMS service relies on SGX. But I add a readme file about how to set up eHSM-KMS service locally. So that anyone can test eHSM-KMS Client on a device with SGX capabilities.

Reply:
Ok. I guess that's why these tests are marked as ignore. You might want to replace your IP with a phony one.